docs/understand/programming_model.rst

```diff
@@ -0,0 +1,158 @@
+SIMD-capable architectures the four-vector ops are expected to compile to a
+single SIMD instruction. GPU execution of this kernel however will typically
+look the following:
+
+.. _simt:
+
+.. figure:: ../data/understand/programming_model/simt.svg
+  :alt: Two large arrows pointing downward with blocks inside and ellipses
+        between the arrows. Inside the arrows the same series of blocks with
+        the following texts inside from top to bottom: "ADD", "DIV", "FMA",
+        "FMA", "FMA" and "FMA".
+
+  Instruction flow of the sample SIMT program.
+
+In HIP, lanes of a SIMD architecture are fed by mapping threads of a SIMT
+execution, one thread down each lane of a SIMD engine. Execution parallelism
+isn't exploited from the width of the built-in vector types, but via the thread
+id constants ``threadIdx.x``, ``blockIdx.x``, etc. For more details, refer to
+:ref:`inherent_thread_model`.
+
+Heterogenous Programming
+===============================================================================
+
+The HIP programming model assumes two execution contexts. One is referred to as
+*host* while compute kernels execute on a *device*. These contexts have
+different capabilities, therefor slightly different rules apply. The *host*
+execution is defined by the C++ abstract machine, while *device* execution
+follows the HIP model, primarily defined by SIMT. These execution contexts in
+code are signified by the ``__host__`` and ``__device__`` decorators. There are
+a few key differences between the two:
+
+* The C++ abstract machine assumes a unified memory address space, meaning that
+  one can always access any given address in memory (assuming the absence of
+  data races). HIP however introduces several memory namespaces, an address
+  from one means nothing in another. Moreover not all address spaces are
+  accessible from all contexts.
+
+  If one were to look at {ref}`cdna2_gcd` and inside the {ref}`cdna3_cu`,
+  every Compute Unit has an instance of storage backing the namespace
+  ``__shared__``. Even if the host were to have access to these regions of
+  memory, the performance benefits of the segmented memory subsystem are
+  supported by the inability of meaningful asynchronous accesss from the host.
+
+* Not all C++ language features map cleanly to typical device architectures,
+  some are very expensive (meaning: slow) to implement on GPU devices, therefor
+  they are forbidden in device contexts to avoid users tapping into features
+  unexpectedly decimating their program's performance. Offload devices targeted
+  by HIP aren't general purpose devices, at least not in the sense a CPU is.
+  HIP focuses on data parallel computations and as such caters to throughput
+  optimized architectures, such as GPUs or accelerators derived from GPU
+  architectures.
+
+* Asynchrony is at the forefront of the HIP API. Computations launched by HIP
+  execute asynchronously on the device and it is the user's responsibility to
```

Comment:
```suggestion
* Asynchrony is at the forefront of the HIP API. Computations launched on the device
  execute asynchronously with respect to the host and it is the user's responsibility to
```
I think it's important to emphasize, that the execution is not asynchronously on the device tiself